### PR TITLE
[ZEPPELIN-5760] fix passing configs to spark in k8s

### PR DESCRIPTION
**Pull request by zlosim**
### What is this PR for?
Some important spark configs can`t be passed via SparkConf in k8s and in client mode including spark.jars.packages and spark.driver.extraJavaOptions
This fix checks if these are set in interpreter configuration and passes them to spark-submit command 

### What type of PR is it?
Bug Fix

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-5760

### How should this be tested?
updated test to test this fix

### Questions:
* Does the licenses files need to update? No
* Is there breaking changes for older versions?  No
* Does this needs documentation? No

### CODE REVIEW

**Comment by Reamer** [comment]
I think most Spark configuration values can be included in the environment variable `SPARK_SUBMIT_OPTIONS`. Please note that the K8sProcess works for all Zeppelin interpreters.

**Comment by zlosim** [comment]
Hi @Reamer  I'm sorry but I am not sure if I understand you correctly. 
As you stated, spark configs can be included in the `SPARK_SUBMIT_OPTIONS ` but they were NOT. This fix specifically adds these configs to `SPARK_SUBMIT_OPTIONS`
there are also two spark configs that CAN'T be passed in `SPARK_SUBMIT_OPTIONS` because of some functionality happening in 'interpreter.sh' thus they are passed in 'SPARK_DRIVER_EXTRAJAVAOPTIONS_CONF' 

also this fix only changes behavior for spark only

**Comment by Reamer** [comment]
I understand your motivation. I just don't like the list of Spark configuration at this point, because it can become very long. Spark has very many configuration options and we can't map them all again here in Zeppelin. I am looking for a more generic approach.
@zjffdu  If `SPARK_SUBMIT_OPTIONS` is deprecated, how can we fill in the function?

**Comment by zjffdu** [comment]
@zlosim [SparkInterpreterLauncher](https://github.com/apache/zeppelin/blob/c4c580a37fde649553d336984a94bcb1b2821201/zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/launcher/SparkInterpreterLauncher.java#L80) supports all the spark configurations, I think we can do similar thing in `K8sRemoteInterpreterProcess`

**Comment by zjffdu** [comment]
@zlosim [SparkInterpreterLauncher](https://github.com/apache/zeppelin/blob/c4c580a37fde649553d336984a94bcb1b2821201/zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/launcher/SparkInterpreterLauncher.java#L80) supports all the spark configurations. It detects all the spark configuration and concatenate them via env variable `ZEPPELIN_SPARK_CONF`.
I think we can do similar thing in `K8sRemoteInterpreterProcess`.

The reason why we deprecate `SPARK_SUBMIT_OPTIONS` is we would like to encourage user to configure spark in Zeppelin via a unified approach like spark (just configure each property one by one), otherwise it may cause inconsistency (e.g. Zeppelin admin may configure `spark.driver.memory` via `SPARK_SUBMIT_OPTIONS` in `zeppelin-env.sh`,  but another user also define it via inline configuration, It is hard for Zeppelin to detect this kind of conflicts).  So even for `ZEPPELIN_SPARK_CONF` in `SparkInterpreterLauncher`, it is for internal usage, it is not exposed to zeppelin users.

**Comment by zlosim** [comment]
Hi @zjffdu thanks for the response.
To be honest, I think this fix is doing exactly what you are saying you want to achieve. Right now setting e.g. [spark dependencies via zeppelin interpreter settings](https://i.imgur.com/VeSvIFa.png) is not working in K8s because we are trying to set it in runtime which is too late as driver is already started. This PR picks settings from interpreter settings passed by user that have to be passed directly to spark submit command when creating k8s pod with driver and puts them in env var so interpreter.sh script will start driver correctly. For end user this is what we are trying to achieve - he sets his settings via zeppelin interepreter settings and everything works as expected. 
As for `SPARK_SUBMIT_OPTIONS` and possibility of zeppelin admin alter configs - [we can alter the append and ignore value](https://github.com/apache/zeppelin/blob/ed251f4b48215f372f437f4490d7f07680793b7b/zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java#L324) effectively leaving `SPARK_SUBMIT_OPTIONS` unexposed to zeppelin users - same as `ZEPPELIN_SPARK_CONF`.

**Comment by zjffdu** [comment]
@zlosim How about updating [buildSparkSubmitOptions](https://github.com/apache/zeppelin/blob/master/zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java#L406)
to what we did in [SparkInterpreterLauncher](https://github.com/apache/zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/launcher/SparkInterpreterLauncher.java#L203). Just collect all the spark properties (start with prefix `spark.`) and concatenate them into `ZEPPELIN_SPARK_CONF` and pass it to `interpreter.sh`, instead of collecting each individual spark driver property explicitly.

**Comment by zlosim** [comment]
Hi @zjffdu , sorry for the late response. I rewrote it so we are now using ZEPPELIN_SPARK_CONF. Please let me know if there is anything I can do better.

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (jongyoul) [inline zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:24]: nit. Could you please change your IDE setting not to use `*`?

Comment (zlosim) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java]: good point :) fixed

**Review by Reamer — CHANGES_REQUESTED** [review]
I think I have found a bug. Have you tested the changes locally?

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (zlosim) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java]: Hi @Reamer, you are right, got my builds messed up when testing, sorry about that. I have just rebuild everything from scratch and tested everything to confirm everything works as expected

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (Reamer) [inline zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:435]: Why `""` at the end?

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (jongyoul) [inline zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:333]: nit. You can use `StringJoiner(" ")` to reduce `.append(" ");`

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (jongyoul) [inline zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:425]: Sorry for the fundamental question, but I wonder if `|` is correct?

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (jongyoul) [inline zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:341]:
```suggestion
        addSparkK8sProperties();
```

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (Reamer) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:425]: Yes `|` is correct. Take a look into https://github.com/apache/zeppelin/blob/4e9cab538e75faa9f4168143f1fdd25c27abedb2/bin/interpreter.sh#L259
and the discussion https://github.com/apache/zeppelin/pull/4127.

Comment (jongyoul) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:425]: Aha ~ Thank you!

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (zlosim) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java]: fixed

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (zlosim) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java]: fixed

**Comment by zlosim** [comment]
Hi @zjffdu , I think this PR is ready to merge unless there are any other requests

**Comment by Reamer** [comment]
I would like to test this beforehand in my local K8s cluster, but I am still on vacation right now. Please have a little patience.

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (Reamer) [inline zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:427]: Possible error.
This creates "--proxy-user|user". desired is "--proxy-user|user"

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (Reamer) [inline zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:433]: Here the same error.
desired is "--conf spark=123|--conf spark.1=123"

**Review by Reamer — COMMENTED** [review]
I think I have found a mistake. Let me check this. The purpose of the pipe is that the individual config values can be cleanly separated from other config values. A separation in the config values itself is not desired.

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (zlosim) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:427]: Hi @Reamer I`m not sure I can see the bug here, right now the output of this snippet is `--proxy-user|user` and the pipe would be replaced in interpreter.sh with space
as far as i can see same code snippet can be found in [SparkInterpreterLauncher](https://github.com/apache/zeppelin/blob/c4c580a37fde649553d336984a94bcb1b2821201/zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/launcher/SparkInterpreterLauncher.java#L196)

File: zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java
Comment (zlosim) [reply zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:433]: similar to the previous one, in [SparkInterpreterLauncher](https://github.com/apache/zeppelin/blob/c4c580a37fde649553d336984a94bcb1b2821201/zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/launcher/SparkInterpreterLauncher.java#L203) desired output is `--conf|spark=123|--conf|spark.1=123` but can fix it to 
`--conf spark=123|--conf spark.1=123` if that is really what we want, please let me know

**Comment by Reamer** [comment]
Hi @zlosim 
Sorry for the late reply. I tried to test your branch. Currently, I am getting a startup error because I have already fully transitioned my environment to Spark-3.3. Could you please do a rebase to the master.

**Comment by zlosim** [comment]
Hi @Reamer 
no prob, should be rebased

**Review by Reamer — APPROVED** [review]
LGTM.
Thank you for your pull request, with this it is possible to remove the deprecated SPARK_SUBMIT option in the K8s launcher.
I have successfully tested your patch in my cluster.
As indicated, it might cause problems in the future if the pipe is also between "--conf" and the attachment. But since this is also done in the other start command, this should not block the merge.

**Pull request merged**
